File: src/main/java/net/hoz/netproto/TimeUtils.java
package net.hoz.netproto;

import com.google.protobuf.Timestamp;
import lombok.experimental.UtilityClass;

import java.time.Instant;

/**
 * Utility class for converting {@link Instant} into {@link Timestamp} and vice-versa.
 */
@UtilityClass
public class TimeUtils {

    public Instant convert(Timestamp timestamp) {
        return Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
    }

    public Timestamp convert(Instant instant) {
        return Timestamp.newBuilder()
                .setSeconds(instant.getEpochSecond())
                .setNanos(instant.getNano())
                .build();
    }
}

File: src/main/kotlin/net/hoz/netproto/TimeUtils.kt
package net.hoz.netproto

import com.google.protobuf.Timestamp
import java.time.Instant

fun Instant.asProto() = Timestamp.newBuilder()
    .setSeconds(this.epochSecond)
    .setNanos(this.nano)
    .build()

fun Timestamp.asInstant() = Instant.ofEpochSecond(this.seconds, this.nanos.toLong())